### PR TITLE
Integrate DeepSeek Sparse Attention with Tokamax Flash Attention

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

Integrate DSA with Tokamax Flash Attention
* Leverages `make_dynamic_splash_mha` to support 3D dynamic index masking within the Flash Attention
* Moves index_mask reshaping to occur locally within the corresponding dot_product attention call

# Tests

* Added dedicated unit tests for the Flash Attention path. Note that because it requires seq must be a multiple of 128 (TPU num of lanes), new test sequence length has been standardized to 128. Observed that **both dot_product and flash attention** require normalization to pass at longer sequence length like seq=128.
  * Add unit tests and run with `python3 -m unittest tests.unit.deepseek32_vs_reference_test` - logs: [link](https://paste.googleplex.com/6085075034832896)
  * Performed a manual diff of the final three entries of the last batch. I do observed 1-2 outliers with index_topk=4 between the dot_product and flash_attention implementations - [link](https://screenshot.googleplex.com/39H45VwRKWbEYBE)
  * Verified that the batch mask indices are also identical across both attention paths - [link](https://diff.googleplex.com/#key=ap6vItMoxAyH)
  * Also tested with a different topK value on the reference side resulted in unit test failures for both dot_product and flash_attention paths, even when normalization was applied. Given that the failures are consistent across both implementations, utilizing normalization within the unit tests appears to be an acceptable baseline for stability.
* Observed a ~10% throughput boost (from 74 tflops/s/device to 82 tflops/s/device) compared to the baseline implementation, tested on a small model version of DS v3.2 (2B model size). More benchmarks will be covered in b/469549024.

```
V3.2 version (2B) - batch=12, seq=4096, FSDP

Per train step:
 Total TFLOPs: 145.89 

# dot_product
I0204 22:40:21.182972 140619320139328 metric_logger.py:179] completed step: 9, seconds: 1.951, TFLOP/s/device: 74.797, Tokens/s/device: 12599.730, total_weights: 98304, loss: 11.982

# flash 
I0204 22:45:41.304875 139758854504000 metric_logger.py:179] completed step: 9, seconds: 1.765, TFLOP/s/device: 82.661, Tokens/s/device: 13924.521, total_weights: 98304, loss: 11.980
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/layers/attention_op.py [hunk 1]
```diff
@@ -1297,10 +1305,25 @@ def wrap_flash_attention(
...
-        )
+        if self.config.use_sparse_indexer and index_mask is not None:
+          # Construct the splash kernel call with dynamic mask
+          def dynamic_mask_splash_kernel(q, k, v, segment, sinks, mask):
+            splash_kernel = tokamax_splash_kernel.make_dynamic_splash_mha(
+                mask=mask,
+                config=sa_config,
+            )
+            kernel = partial(splash_kernel, max_logit_value=max_logit_value)
+            return kernel(q, k, v, segment, sinks)
```
Comment (shuningjin) [inline src/MaxText/layers/attention_op.py]: Does sinks need to be kwarg here? I see  `else` branch has `kernel(q, k, v, d, sinks=s)`

**Review by shuningjin — COMMENTED** [review]
Thank you for the tokamax flash integration! Excited to see throughput improving! Overall looks good.

---

> Observed that both dot_product and flash attention require normalization to pass at longer sequence length like seq=128.

Yes, normalization is needed with larger seq. In the test, the torch params are manually initialized with N(0, 1); without init they are all zeros.
https://github.com/AI-Hypercomputer/maxtext/blob/95ef3e1074f37126cfd2432b7d0376f24cb82632/tests/unit/deepseek32_vs_reference_test.py#L708-L715
As a result of std=1, the logits are much larger than what we usually have, e.g., 2000 ([log](https://paste.googleplex.com/4905500253749248#l=162)). Also I see normalization is used for gpt-oss attention tests.
https://github.com/AI-Hypercomputer/maxtext/blob/95ef3e1074f37126cfd2432b7d0376f24cb82632/tests/unit/gpt_vs_reference_test.py#L445

---

> Performed a manual diff of the final three entries of the last batch. I do observed 1-2 outliers with index_topk=4 between the dot_product and flash_attention implementations

Do you think we can add a test that directly compare flash against dot product, like what we have in `attention_test.py`?

https://github.com/AI-Hypercomputer/maxtext/blob/95ef3e1074f37126cfd2432b7d0376f24cb82632/tests/unit/attention_test.py#L478-L479

either inside `deepseek32_vs_reference_test.py` (e.g., [this](https://paste.googleplex.com/5931309907181568) can pass) or `attention_test`.

File: src/MaxText/layers/attention_op.py (hunk 1 above)
Comment (RissyRan) [reply src/MaxText/layers/attention_op.py]: Good call, let's just add it for more readable and robust.

File: tests/unit/deepseek32_vs_reference_test.py [hunk 2]
```diff
@@ -907,17 +911,53 @@ class DeepseekV32MLATest(DeepseekTestBase):
...
+          "check_norm": True,
+      },
+      {
+          "testcase_name": "seq_len=128 (index_topk=4) & flash",
+          "attention": "flash",
+          "seq_len": 128,
+          "index_topk": 4,
+          "check_norm": True,
+      },
+      {
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply tests/unit/deepseek32_vs_reference_test.py:946]: Updated.

File: tests/unit/deepseek32_vs_reference_test.py [hunk 3]
```diff
@@ -907,17 +911,53 @@ class DeepseekV32MLATest(DeepseekTestBase):
...
+          "testcase_name": "seq_len=128 (index_topk=128) & flash",
+          "attention": "flash",
+          "seq_len": 128,
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply tests/unit/deepseek32_vs_reference_test.py:949]: Updated.

**Comment by RissyRan** [comment]
> Thank you for the tokamax flash integration! Excited to see throughput improving! Overall looks good.
> 
> > Observed that both dot_product and flash attention require normalization to pass at longer sequence length like seq=128.
> 
> Yes, normalization is needed with larger seq. In the test, the torch params are manually initialized with N(0, 1); without init they are all zeros.
> 
> https://github.com/AI-Hypercomputer/maxtext/blob/95ef3e1074f37126cfd2432b7d0376f24cb82632/tests/unit/deepseek32_vs_reference_test.py#L708-L715
> 
> 
> As a result of std=1, the logits are much larger than what we usually have, e.g., 2000 ([log](https://paste.googleplex.com/4905500253749248#l=162)). Also I see normalization is used for gpt-oss attention tests.
> https://github.com/AI-Hypercomputer/maxtext/blob/95ef3e1074f37126cfd2432b7d0376f24cb82632/tests/unit/gpt_vs_reference_test.py#L445
> 
> > Performed a manual diff of the final three entries of the last batch. I do observed 1-2 outliers with index_topk=4 between the dot_product and flash_attention implementations
> 
> (Optional) Do you think we can add a test that directly compare flash against dot product, like what we have in `attention_test.py`?
> 
> https://github.com/AI-Hypercomputer/maxtext/blob/95ef3e1074f37126cfd2432b7d0376f24cb82632/tests/unit/attention_test.py#L478-L479
> 
> either inside `deepseek32_vs_reference_test.py` (e.g., [this](https://paste.googleplex.com/5931309907181568) can pass) or `attention_test`.

Thanks for giving that a shot, Shuning! If this normalization holds up, I’d actually prefer to compare it directly against the reference implementation. Since the codebase is evolving so fast, there's a risk of regressions in dot_product as well. Comparing against the reference feels like the safest bet—what do you think?

File: src/MaxText/layers/attention_op.py [hunk 4]
```diff
@@ -1051,6 +1054,7 @@ def tpu_flash_attention(
     segment_axis_names_q = None
     segment_axis_names_kv = None
     sink_axis_names = self._logical_to_mesh_axes((HEAD,))
+    index_mask_axis_names = self._logical_to_mesh_axes((BATCH,))
```
Comment (gagika) [inline src/MaxText/layers/attention_op.py]: does it support context parallelism? 
Gemini suggest having something like this to support it:

```
    # Define index_mask sharding to match query sequence partitioning
    if self.config.expert_shard_attention_option == EP_AS_CONTEXT:
        index_mask_axis_names = self._logical_to_mesh_axes((BATCH_NO_EXP, Q_LENGTH, KV_LENGTH))
    else:
        index_mask_axis_names = self._logical_to_mesh_axes((BATCH, Q_LENGTH_NO_EXP, KV_LENGTH))

```

Comment (RissyRan) [reply src/MaxText/layers/attention_op.py]: Good call! I don't think it supports EP_AS_CONTEXT yet. Let me have a try quickly.

Comment (RissyRan) [reply src/MaxText/layers/attention_op.py]: I updated to 

```
    if self.config.expert_shard_attention_option == EP_AS_CONTEXT:
      axis_names_splash_kernel = self._logical_to_mesh_axes(self.flash_axis_names_splash_kernel_ep)
      axis_names_q = self._logical_to_mesh_axes(self.flash_axis_names_q_ep)
      axis_names_kv = self._logical_to_mesh_axes(self.flash_axis_names_kv_ep)
      index_mask_axis_names = self._logical_to_mesh_axes((BATCH_NO_EXP, Q_LENGTH, KV_LENGTH))
    else:
      axis_names_splash_kernel = self._logical_to_mesh_axes(self.flash_axis_names_splash_kernel)
      axis_names_q = self._logical_to_mesh_axes(self.flash_axis_names_q)
      axis_names_kv = self._logical_to_mesh_axes(self.flash_axis_names_kv)
      index_mask_axis_names = self._logical_to_mesh_axes((BATCH, Q_LENGTH, KV_LENGTH))
```

Tried both cases with `ici_expert_parallelism=2`, and working: https://paste.googleplex.com/5991084107890688

Will update accordingly
